File: samsung-sds/boj_3425.py
# 백준 - 고스택

import sys

cmd_list = []
stack = []
answer = []

def calculate(input):
    # print(input)
    global stack

    if input[0:3] == "NUM":
        num = int(input.split(" ")[1])
        stack.append(num)
        return True

    elif input[0:3] == "POP":
        if stack:
            stack.pop()
            return True
        return False

    elif input[0:3] == "INV":
        if stack:
            num = stack.pop()
            stack.append(-num)
            return True
        return False

    elif input[0:3] == "DUP":
        if stack:
            num = stack[-1]
            stack.append(num)
            return True
        return False

    elif input[0:3] == "SWP":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()
            stack.append(num1)
            stack.append(num2)
            return True
        return False

    elif input[0:3] == "ADD":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()
            result = num1 + num2
            if check(result):
                stack.append(result)
                return True
        return False

    elif input[0:3] == "SUB":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()
            result = num2 - num1
            if check(result):
                stack.append(result)
                return True
        return False

    elif input[0:3] == "MUL":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()
            result = num1 * num2
            if check(result):
                stack.append(result)
                return True
        return False

    elif input[0:3] == "DIV":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()

            if num1 == 0:
                return False
            result = abs(num2) // abs(num1)
            # 음수가 하나라도 있을때
            if num1 < 0 or num2 < 0:
                if not (num1 < 0 and num2 < 0):
                    result *= -1

            stack.append(result)
            return True
        return False

    elif input[0:3] == "MOD":
        if len(stack) > 1:
            num1 = stack.pop()
            num2 = stack.pop()  # 피제수

            if num1 == 0:
                return False

            result = abs(num2) % abs(num1)
            # 음수가 하나라도 있을때
            if num2 < 0:
                result *= -1
            stack.append(result)
            return True
        return False

    return False


def check(num):
    MAX_NUM = 10 ** 9
    if abs(num) > MAX_NUM:
        return False
    return True


while True:

    input = sys.stdin.readline()
    if input == '\n':
        cmd_list.clear()
        continue

    input = input.strip()



    if input == "QUIT":
        for str in answer:
            print(str)

        sys.exit(0)


    elif input == "END":
        N = int(sys.stdin.readline())
        for _ in range(N):
            num = int(sys.stdin.readline())
            stack = [num]

            for cmd in cmd_list:
                result = calculate(cmd)
                if not result:
                    answer.append("ERROR")
                    break

            else:
                if len(stack) == 1:
                    answer.append(stack[-1])

                else:
                    answer.append("ERROR")

        answer.append(" ")
        continue
    else:
        cmd_list.append(input)

File: stack/Boj3425.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;

import static java.lang.Math.abs;
import static java.lang.Math.pow;

public class Boj3425 {
    static LinkedList<Long> stack;
    static int stackSize;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        ArrayList<String> cmdList = new ArrayList<>();
        String[] input = br.readLine().split(" ");

        stack = new LinkedList<>();

        while(!input[0].equals("QUIT")) {
            if(input[0].equals("")) {
                cmdList.clear();
                System.out.println();
            }
            else if(input[0].equals("END")) {
                int N = Integer.parseInt(br.readLine().split(" ")[0]);

                for(int n=0; n < N; n++) {
                    stack.clear();
                    long initNum = Long.parseLong(br.readLine().split(" ")[0]);
                    stack.add(initNum);
                    System.out.println(runCommand(cmdList));
                }

            }
            else {
                cmdList.add(input[0]);
                if(input[0].equals("NUM")) {
                    cmdList.add(input[1]);
                }
            }
            input = br.readLine().split(" ");

        }

    }

    private static String runCommand(ArrayList<String> cmd) {
        int idx = 0;
        while (idx < cmd.size()) {
            try {
                switch (cmd.get(idx)) {
                    case "NUM": {
                        long x = Long.parseLong(cmd.get(++idx));
                        stack.add(x);
                        break;
                    }
                    case "POP": {
                        stack.removeLast();
                        break;
                    }
                    case "INV": {
                        long tmp = stack.removeLast();;
                        stack.add(tmp*-1);
                        break;
                    }
                    case "DUP": {
                        stack.add(stack.getLast());
                        break;
                    }
                    case "SWP": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();

                        stack.add(num1);
                        stack.add(num2);
                        break;
                    }
                    case "ADD": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();

                        stack.add(num1+num2);
                        stackSize--;
                        break;
                    }
                    case "SUB": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();

                        stack.add(num2-num1);
                        stackSize--;
                        break;
                    }
                    case "MUL": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();

                        stack.add(num2*num1);
                        stackSize--;
                        break;
                    }
                    case "DIV": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();


                        long tmp = abs(num2)/abs(num1);
                        if(num1*num2 < 0) {
                            tmp *= -1;
                        }
                        stack.add(tmp);
                        break;

                    }
                    case "MOD": {
                        long num1 = stack.removeLast();
                        long num2 = stack.removeLast();


                        long tmp = abs(num2)%abs(num1);
                        if(num2<0) {
                            tmp *= -1;
                        }

                        stack.add(tmp);
                        stackSize--;
                        break;
                    }
                }
            } catch (Exception e) {
                return "ERROR";
            }
            idx++;
        }

        if(stack.size() == 1) {
            if(abs(stack.getLast()) > pow(10, 9)) {
                return "ERROR";
            }
            return stack.getLast().toString();

        }
        return "ERROR";


    }
}
